common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -496,41 +440,116 @@ public void run() {
             } else {
                 //Update Dialog to indicate that an incorrect attempt was made.
-                synchronized (sSmartcardDialogLock) {
-                    ((SmartcardPinDialog) mCurrentDialog).setErrorMode();
-                }
+                mDialogHolder.setPinDialogErrorMode();
             }
         }
     }
 
     /**
-     * Sets up and shows dialog when user has exceeded the max attempts allowed to enter their YubiKey PIN.
+     * Attempts to authenticate using smartcard certificate.
+     * @param cert Chosen certificate for authentication.
+     * @param pin char array containing PIN.
+     * @param slotAlias Name of slot where certificate is located.
+     * @param piv A PivSession created from a SmartCardConnection that can interact with certificates located in the PIV slots on the YubiKey.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
      */
-    private void showMaxAttemptsDialog(){
-        final SmartcardMaxFailedAttemptsDialog maxFailedAttemptsDialog = new SmartcardMaxFailedAttemptsDialog(
-                new SmartcardMaxFailedAttemptsDialog.PositiveButtonListener() {
-                    @Override
-                    public void onClick() {
-                        //Reset currentDialog to null
-                        synchronized (sSmartcardDialogLock) {
-                            mCurrentDialog = null;
-                        }
-                    }
-                },
-                mActivity);
-        //set current dialog to maxFailedAttemptsDialog and show.
-        synchronized (sSmartcardDialogLock) {
-            mCurrentDialog = maxFailedAttemptsDialog;
-            maxFailedAttemptsDialog.show();
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void useSmartcardCertForAuth(@NonNull final X509Certificate cert,
+                                         @NonNull final char[] pin,
+                                         @NonNull final String slotAlias,
+                                         @NonNull final PivSession piv,
+                                         @NonNull final ClientCertRequest request) {
+        final String methodTag = TAG + "useSmartcardCertForAuth:";
+        //Need to add a PivProvider instance to the beginning of the array of Security providers in order for signature logic to occur.
+        //Note that this provider is removed when the UsbYubiKeyDevice connection is closed.
+        //First check if a PivProvider instance is already present in the static list.
+        if (Security.getProvider(YKPIV) != null) {
+            //Remove existing PivProvider.
+            Security.removeProvider(YKPIV);
+            //Telemet and log this, as the PivProvider instance is either unexpectedly being added elsewhere
+            // or it isn't being removed properly upon CBA flow termination.
+            Telemetry.emit((BaseEvent) new BaseEvent().put(
```

Comment:
I would suggest emitting "false" also when the provider is not there - so that we can calculate the percentage.